### PR TITLE
Autoconnect ALSA MIDI inputs.

### LINKED ISSUE #156: Auto-connect ALSA MIDI inputs
While using QSynth+fluidsynth mostly as a hobbist, I came across this need for
reconnecting the ALSA MIDI input to QSynth/fluidsynth everytime I restart the engine.

So I worked out the attached patch that adds a bool setting for auto-connecting ALSA
MIDI inputs. This is the dual of the "Auto-connect JACK outputs" already there in
fluidsynth/QSynth.

If this makes sense, I'd follow up with a corresponding patch to QSynth to set this
property when instantiating the fluidsynth engine.

Reported by: tcucinotta

Original Ticket: [fluidsynth/tickets/158](https://sourceforge.net/p/fluidsynth/tickets/158)

### PR DESCRIPTION
**Pull request by derselbst**
fixes #156

### CODE REVIEW

**Comment by derselbst** [comment]
@tomcucinotta FYI

**Comment by tomcucinotta** [comment]
hi there, thanks for the notification, can I do anything to help out here ? my old patch was from ~1year ago, was it rebased ?

File: doc/fluidsettings.xml [hunk 1]
```diff
@@ -240,6 +240,14 @@ https://stackoverflow.com/a/6251757
             </desc>
         </setting>
         <setting>
+            <name>alsa.autoconnect</name>
+            <type>bool</type>
+            <def>0 (FALSE)</def>
+            <desc>
+                If 1 (TRUE), automatically connects FluidSynth to available ALSA MIDI input ports.
```
Comment (derselbst) [inline doc/fluidsettings.xml]: @tomcucinotta Haven't tested it yet. You could let me know if this documentation is sufficient.

Comment (tomcucinotta) [reply doc/fluidsettings.xml]: docs seem ok, thanks!

Comment (tomcucinotta) [reply doc/fluidsettings.xml]: Hi, I also just recompiled this new alsa-autocon branch, and tried the feature, plugging my USB MIDI adapter, then launching:

  ./src/fluidsynth -o audio.alsa.autoconnect=1 -a alsa -i /usr/share/sounds/sf2/FluidR3_GM.sf2 -s

I've got the USB MIDI adapter, as well as the MIDI through, readily connected to FLUID Synth :-)!

Comment (derselbst) [reply doc/fluidsettings.xml]: Confirmed.

**Comment by mawe42** [comment]
> But sending data to those ports doesn't work, of course, because they are not sequencer but raw ports.
Forget that. Sending data does work, just not with aplaymidi. If I send raw midi events to the port, FluidSynth reacts just fine.

So the only remaining valid comment is the suggestion to rename `audio.alsa.autoconnect` to `midi.alsa.autoconnect`.

**Comment by derselbst** [comment]
Ok, but should it be `midi.alsa.autoconnect` or `midi.alsa_seq.autoconnect`? It applies for both. I would also vote for the first. Or we have to make it work only with real alsa_seq ports, not with raw ones.

**Comment by tomcucinotta** [comment]
anyone willing to explain what a raw MIDI port is useful for? I'm just reading from alsa docs that it doesn't add timestamps, but I confess I don't really know what it's talking about ... sorry for the newbie Q, thanks.

**Comment by derselbst** [comment]
Never figured that out myself. All I know: "There are seq ports and raw ports. Always use seq ports!" Sry.

**Comment by mawe42** [comment]
A raw MIDI port is basically what your USB keyboard creates when it attaches to your computer. It's a port that outputs (or takes in) raw MIDI events. So no MIDI files, with headers etc, just the plain MIDI protocol data in binary form. Attaching FluidSynth with alsa_raw and midi.alsa.device=hw:1,0 (or similar) to your MIDI device is the best way for very low latency MIDI input. Going through alsa_seq always adds a little more latency, as I understand it.

**Comment by mawe42** [comment]
Also, alsa_raw ports are usually opened in exclusive mode, so only one application can use the port. The sequencer ports can be subscribed to by many clients at the same time. But like I said, this flexibility comes at the cost of higher resource consumption and latency.

**Comment by mawe42** [comment]
And with regard to the timestamps in alsa_seq, I think it works as follows: a raw midi event (like note on, channel 0, note 60, velocity 127) get sent by the MIDI device (for example a keyboard). It comes in via a raw midi device and gets processed by the alsa sequencer engine. The sequencer adds a timestamp to it saying "this is when I received the event". Then the sequencer passes on the event with the timestamp to all connected clients.

For real time control, those timestamps would not be necessary. But the alsa_seq interface can also take events with a timestamp in the future. And that is basically what a MIDI file is.

**Comment by derselbst** [comment]
Aha, so does it even make sense then connecting to raw ports?

**Comment by mawe42** [comment]
> Aha, so does it even make sense then connecting to raw ports?

Yes, very much so. Especially if you want to control FluidSynth in real time with a MIDI instrument or other MIDI controller.

**Comment by derselbst** [comment]
Ok, so `midi.alsa.autoconnect` or `midi.alsa_seq.autoconnect`?

**Comment by mawe42** [comment]
With the current implementation, it should probably be `midi.alsa_seq.autoconnect`, because it uses the ALSA sequencer api and doesn't have any effect for `-m alsa_raw`.

Or we could go for `midi.autoconnect`, assuming other midi drivers could also benefit from an auto-connect feature. But I don't know anything about coremidi or Windows MIDI...

**Pull request merged**